timemachine/ff/smirnoff_converter.py

```diff
@@ -84,6 +90,106 @@ def parse_quantity(number_string):
+        Equilibrium value in rad.
+
+    cos_angles:
+        Set to True to return the cos angles version of the angle potential energy.
+        Otherwise return the standard form.
+    """
+    if cos_angles:
+        return 0.5 * k * (cos_theta - jnp.cos(theta_0)) ** 2
+    else:
+        return 0.5 * k * (jnp.arccos(cos_theta) - theta_0) ** 2
+
+
+def get_pdf(x_grid: NDArray, k: float, theta_0: float, cos_angles: bool) -> NDArray:
+    """
+    Return the normalized probability as a function of cos(theta).
+
+    Parameters
+    ----------
+    x_grid:
+        Grid from -1 to +1. Should be fine to accuately compute the normalization constant.
+
+    k:
+        Force constant in (kJ/mol)/rad**2
+
+    theta_0:
+        Equilibrium value in rad.
+
+    cos_angles:
+        Set to True to return the cos angles version of the angle potential energy.
+        Otherwise return the standard form.
+    """
+    q_ref = jnp.exp(-U_angle_cos(x_grid, k=k, theta_0=theta_0, cos_angles=cos_angles)) / DEFAULT_KT
+    Z_ref = jnp.trapz(q_ref, x_grid)
+    return q_ref / Z_ref
+
+
+def kl_divergence(x_grid: NDArray, pdf_a: NDArray, pdf_b: NDArray) -> float:
+    """
+    Return the Kullback-:eibler divergence between two pdfs.
+    NOTE: This is asymmetric, d(a, b) != d(b, a).
+    """
+    integrand = jnp.nan_to_num(pdf_a * (jnp.log(pdf_a) - jnp.log(pdf_b)), nan=0)
+    return jnp.trapz(integrand, x_grid)
+
+
+def js_divergence(x_grid: NDArray, pdf_a: NDArray, pdf_b: NDArray) -> float:
+    """
+    Return the Jensen–Shannon divergence between the two pdfs.
+
+    NOTE: This is symmetric, d(a, b) = d(b, a).
+    """
+    pdf_mid = 0.5 * (pdf_a + pdf_b)
+    return 0.5 * (kl_divergence(x_grid, pdf_a, pdf_mid) + kl_divergence(x_grid, pdf_b, pdf_mid))
+
+
+def make_cos_angle_loss_fxn(x_grid: NDArray, k: float, theta_0: float) -> Callable[[float], float]:
+    """
+    Return a loss function used to minimize the JS divergence bewteen the
+    original and cos angle pdfs.
+    """
+    pdf_ref = get_pdf(x_grid, k, theta_0, cos_angles=False)
+
+    def loss_fxn(k_prime):
+        pdf_test = get_pdf(x_grid, k_prime, theta_0, cos_angles=True)
+        distance = js_divergence(x_grid, pdf_ref, pdf_test)
+        return distance
+
+    return loss_fxn
+
+
+def scale_angle_params(k: float, a0: float) -> float:
+    """
+    Scale the force constant for the angle term to account for the cos angle
+    fucntional form: V(t) = k*(cos(t)-cos(a0))^2 compared to the original functional
+    form V(t) = k*(t - a0)^2.
+
+    Based on jfass's optimization code.
+    """
+    # Vacuum simulation of propyne is unstable
+    # if the angle k is above 12k, set below for some buffer.
+    MAX_ANGLE_K = 10000.0  # (kJ/mol)/rad**2
+    x_grid = np.linspace(-1, +1, 10000, dtype=np.float64)
+    loss_fxn = make_cos_angle_loss_fxn(x_grid, k, a0)
+    optimize_result = minimize_scalar(loss_fxn, method="bounded", bounds=(0, MAX_ANGLE_K))
+    return float(optimize_result.x)
```

Comment:
Question: How should this change be validated?

In the version of this script I wrote on GitLab last year, I had just observed that the JS divergence on this 1D surrogate was highest for equilibrium angles near pi

<img width="600" alt="image" src="https://user-images.githubusercontent.com/5759036/222580576-86649aa3-366c-4171-8ceb-2246a22c1840.png">

and observed that optimizing this loss increased many of the angle force constants

<img width="600" alt="image" src="https://user-images.githubusercontent.com/5759036/222579092-14cef242-60a9-47ac-8531-7d346a728466.png">

but (1) some of the minimizations NaN'd (I think this is addressed in the current PR by the addition of bounds), and (2) I did not know how to validate the changes.

Other questions:
* Would it make sense to apply this conversion only to angle parameters where `theta_0` is near `pi`? Currently applied to all angle parameters.
* Would it make sense to add a penalty that keeps the adjusted parameter close to its original value? Currently no penalty.
* Any function of the 1D profiles would be equally convenient to optimize here -- JS divergence was picked intuitively. Is JS divergence a good reflection of our intent in adjusting these parameters?

Reply:
Thanks for walking me through the validation offline!

(It's a compelling large improvement, esp. at theta_0 = pi, and doesn't seem to worsen any other cases. This looks good to me to be merged as-is. However, was surprised that there's still a difference between cos_angles=True and cos_angles=False in some cases where theta_0 < pi. Will postpone further review while you're looking into other possible approaches for those cases, possibly including an alternative that doesn't require converting the parameters.)